### ISSUE TITLE
[Bug] tf.where returns None gradient for differentiable expression

### ISSUE THREAD

**Issue by amadhan882** [tensorflow/tensorflow#115116]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

source

### TensorFlow version

2.19.0

### Custom code

Yes

### OS platform and distribution

Linux-6.6.113+-x86_64-with-glibc2.35 (Google Colab)

### Python version

3.12.13

### CUDA/cuDNN version

No GPU (CPU-only environment)

### Current behavior?

Using `tf.where` with differentiable branches results in a `None` gradient, even though the function is mathematically differentiable.

## Observed Output
```
y: [1. 2. 9.]
grad: None
```
The gradient is not computed at all (None), leading to silent failure.

### Standalone code to reproduce the issue

```shell
import tensorflow as tf

x = tf.Variable([-1.0, 2.0, -3.0])

with tf.GradientTape() as tape:
    y = tf.where(x > 0, x, x * x)

grad = tape.gradient(tf.reduce_sum(y), x)

print("y:", y.numpy())
print("grad:", grad)
```

### Relevant log output

```shell
No error or warning is raised. Gradient is silently None.
```

**Comment by Whatsonyourmind** [comment]
This is expected behavior, though arguably it should be documented more clearly. `tf.where` selects between two pre-computed tensors -- it does not construct a piecewise-differentiable function.

The issue: `tf.where(cond, x, x*x)` evaluates **both** branches before selecting. TensorFlow's autodiff sees the gradient of the full expression, but `tf.where`'s gradient implementation requires gradients from both branches, and the way it back-propagates through the selection makes it return `None` in this case.

The fix is to use `tf.cond` or manual masking:

```python
x = tf.Variable([-1.0, 2.0, -3.0])
with tf.GradientTape() as tape:
    mask = tf.cast(x > 0, tf.float32)
    y = mask * x + (1.0 - mask) * (x * x)

grad = tape.gradient(tf.reduce_sum(y), x)
# grad: [-2.0, 1.0, -6.0]  (correct: d/dx(x^2) = 2x for negative, d/dx(x) = 1 for positive)
```

Mathematically, the piecewise function f(x) = x if x > 0 else x^2 has well-defined gradients everywhere (f'(x) = 1 for x > 0, f'(x) = 2x for x < 0). The masking approach correctly computes this because the gradient flows through both paths weighted by the mask, which is treated as a constant (no gradient through the comparison).

In PyTorch, `torch.where` handles this correctly because it registers custom backward rules for the selection op.

**Comment by cybertarr-A** [comment]
This is not a TensorFlow bug, but rather a limitation of how `tf.where` interacts with automatic differentiation when the condition depends on the same variable being differentiated.

### Root Cause

`tf.where(condition, a, b)` acts as an element-wise selector, not true control flow. When the condition (`x > 0`) depends on `x`, it introduces a non-differentiable boolean mask. As a result, TensorFlow may fail to construct a valid gradient path and return `None` instead of a tensor.

This is a known behavior in autodiff systems when combining:

* boolean masking (`x > 0`)
* branch selection (`tf.where`)
* differentiation w.r.t. the same variable (`x`)

---

### Recommended Fix

Instead of using `tf.where` directly, rewrite the expression using arithmetic masking, which preserves gradient flow:

```python
import tensorflow as tf

x = tf.Variable([-1.0, 2.0, -3.0])

with tf.GradientTape() as tape:
    mask = tf.cast(x > 0, tf.float32)
    y = mask * x + (1 - mask) * (x * x)

grad = tape.gradient(tf.reduce_sum(y), x)

print("y:", y.numpy())
print("grad:", grad.numpy())
```

### Output

```
y: [1. 2. 9.]
grad: [-2.  1. -6.]
```

---

### Alternative (Explicit Gradient Definition)

For full control, you can define a custom gradient:

```python
@tf.custom_gradient
def safe_where(x):
    y = tf.where(x > 0, x, x * x)

    def grad(dy):
        return dy * tf.where(x > 0, 1.0, 2.0 * x)

    return y, grad
```

---

### Summary

* The function is mathematically differentiable, but the computational graph is not smoothly differentiable due to the boolean condition.
* This leads TensorFlow to return `None` instead of a partial gradient.
* Rewriting using arithmetic masking or custom gradients resolves the issue.

---

It might be helpful if TensorFlow emitted a warning in such cases, since the silent `None` gradient can be confusing.
